File: Advanced Robotics Course/Chapter18/Lesson3/Chapter18_Lesson3.java
import java.util.Random;

public class StressTest {

    private static final Random rng = new Random(42L);

    // Sample stress scenarios (simplified distributions).
    static double[] sampleScenario() {
        double friction = Math.max(0.0, Math.min(1.0, rng.nextGaussian() * 0.2 + 0.3));
        double mass = Math.exp(rng.nextGaussian() * 0.25);  // log-normal-ish
        double latency = -0.08 * Math.log(1.0 - rng.nextDouble());
        return new double[]{friction, mass, latency};
    }

    static class TrialResult {
        boolean failed;
        String label;
        TrialResult(boolean f, String l) { failed = f; label = l; }
    }

    static TrialResult simulateTrial(double friction, double mass, double latency) {
        boolean stoppingViolation = (friction * mass) < 0.6;
        boolean latencyViolation = latency > 0.15;
        boolean timeoutViolation = (mass > 2.5) && (latency > 0.10);

        if (stoppingViolation && latencyViolation) {
            return new TrialResult(true, "control+environment:multi-factor");
        }
        if (stoppingViolation) {
            return new TrialResult(true, "control:insufficient_braking");
        }
        if (latencyViolation) {
            return new TrialResult(true, "control:latency_instability");
        }
        if (timeoutViolation) {
            return new TrialResult(true, "performance:timeout");
        }
        return new TrialResult(false, "success");
    }

    public static void main(String[] args) {
        int N = 5000;
        double delta = 0.05;
        int failures = 0;

        for (int i = 0; i < N; ++i) {
            double[] s = sampleScenario();
            TrialResult tr = simulateTrial(s[0], s[1], s[2]);
            if (tr.failed) {
                failures++;
                // In practice, increment a Map<String,Integer> of taxonomy counts.
            }
        }

        double pHat = (double) failures / N;
        double eps = Math.sqrt(Math.log(2.0 / delta) / (2.0 * N));
        double ciLow = Math.max(0.0, pHat - eps);
        double ciHigh = Math.min(1.0, pHat + eps);

        System.out.println("N = " + N);
        System.out.println("Estimated failure probability: " + pHat);
        System.out.println("Hoeffding 95% CI: [" + ciLow + ", " + ciHigh + "]");
    }
}
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter18/Lesson3/Chapter18_Lesson3.py
import numpy as np

# Example: stress scenarios over friction, payload mass, and control latency.
# In practice, replace simulate_trial(...) with calls into PyBullet, MoveIt, etc.

rng = np.random.default_rng(seed=42)

def sample_stress_scenarios(N):
    """Draw N stress scenarios.
    friction: biased toward low values using a Beta distribution.
    mass: log-normal (kg).
    latency: exponential tail (s).
    """
    friction = rng.beta(a=0.5, b=2.0, size=N)        # more mass near 0
    mass = rng.lognormal(mean=0.0, sigma=0.25, size=N)
    latency = rng.exponential(scale=0.08, size=N)    # mean 80 ms
    return friction, mass, latency

def simulate_trial(friction, mass, latency):
    """Return (failed, taxonomy_label).

    failed: bool
    taxonomy_label: string representing failure class, e.g.
      "control:overshoot", "planning:infeasible", "sensing:occlusion", ...
    Here we implement a toy algebraic model instead of a real simulator.
    """
    # Constraint 1: stopping distance constraint violated if
    # effective friction * mass is too small.
    stopping_violation = (friction * mass) < 0.6

    # Constraint 2: closed-loop stability degrades if latency is too large.
    latency_violation = latency > 0.15  # seconds

    if stopping_violation and latency_violation:
        return True, "control+environment:multi-factor"
    if stopping_violation:
        return True, "control:insufficient_braking"
    if latency_violation:
        return True, "control:latency_instability"

    # Example of a non-safety failure: task time-out.
    timeout_violation = (mass > 2.5) and (latency > 0.10)
    if timeout_violation:
        return True, "performance:timeout"

    return False, "success"

def run_stress_test(N, delta=0.05):
    friction, mass, latency = sample_stress_scenarios(N)

    failures = 0
    labels = []
    for f, m, L in zip(friction, mass, latency):
        failed, label = simulate_trial(f, m, L)
        if failed:
            failures += 1
            labels.append(label)

    p_hat = failures / N
    # Hoeffding half-width for the two-sided confidence interval
    eps = np.sqrt(np.log(2.0 / delta) / (2.0 * N))
    ci_low = max(0.0, p_hat - eps)
    ci_high = min(1.0, p_hat + eps)

    # Empirical taxonomy distribution on failures
    from collections import Counter
    counts = Counter(labels)
    tax_probs = {k: v / failures for k, v in counts.items()} if failures > 0 else {}

    return {
        "N": N,
        "failures": failures,
        "p_hat": p_hat,
        "ci": (ci_low, ci_high),
        "taxonomy_distribution": tax_probs,
    }

if __name__ == "__main__":
    result = run_stress_test(N=5000, delta=0.05)
    print("N:", result["N"])
    print("Estimated failure probability:", result["p_hat"])
    print("95% Hoeffding CI:", result["ci"])
    print("Failure taxonomy distribution:")
    for k, v in result["taxonomy_distribution"].items():
        print(f"  {k}: {v:.3f}")
      
